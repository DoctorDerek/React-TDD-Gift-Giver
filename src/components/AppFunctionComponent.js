import React, { useState } from "react"

/**
 * Function component version of the App using React Hooks
 *
 * @param {*} props
 * @return {*}
 */
const AppFunctionComponent = (props) => {
  const [gifts, setGifts] = useState([])

  const addGift = () => {
    const ids = gifts.map((gift) => gift.id)
    // Find max ID; could use Math.max() and ... spread operator
    // But a for loop is faster with large data sets typically
    let maxID = 0
    for (const id of ids) {
      if (id > maxID) {
        maxID = id
      }
    }
    setGifts((gifts) => [...gifts, { id: maxID + 1 }])
  }

  return (
    <div>
      <h2>Gift Giver</h2>
      <button id="add-gift-button" onClick={() => addGift()}>
        Add Gift
      </button>
      <div id="gift-list">
        {gifts.map((gift) => (
          <div key={gift.id}>{gift.id}</div>
        ))}
      </div>
    </div>
  )
}

export default AppFunctionComponent
