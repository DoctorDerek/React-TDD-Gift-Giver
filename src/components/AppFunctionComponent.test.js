import React from "react"
import { shallow } from "enzyme"
import AppFunctionComponent from "./AppFunctionComponent"

let app = shallow(<AppFunctionComponent />)

beforeEach(() => {
  // Clean up between tests:
  app = shallow(<AppFunctionComponent />)
})

// It Renders.
it("Renders correctly", () => {
  expect(app).toMatchSnapshot()
})

describe("when clicking the `Add Gift` button", () => {
  beforeEach(() => {
    app.find("#add-gift-button").simulate("click")
  })

  it("Adds a new gift to the rendered list", () => {
    expect(app.find("#gift-list").children().length).toEqual(1)
  })
})
