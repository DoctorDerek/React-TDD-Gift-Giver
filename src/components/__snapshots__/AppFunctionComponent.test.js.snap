// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppFunctionComponent />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Gift Giver
</h2>,
        <button
          id="add-gift-button"
          onClick={[Function]}
>
          Add Gift
</button>,
        <div
          id="gift-list"
>
          
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Gift Giver",
        },
        "ref": null,
        "rendered": "Gift Giver",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Add Gift",
          "id": "add-gift-button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Add Gift",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "id": "gift-list",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Gift Giver
</h2>,
          <button
            id="add-gift-button"
            onClick={[Function]}
>
            Add Gift
</button>,
          <div
            id="gift-list"
>
            
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Gift Giver",
          },
          "ref": null,
          "rendered": "Gift Giver",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add Gift",
            "id": "add-gift-button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Add Gift",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "id": "gift-list",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
